File: code-external/kmeans.py
import random

def euclidean_distance(a, b):
    return ((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2) ** 0.5

def kmeans(data, k, max_iters):
    # Initialize centroids randomly
    centroids = [(random.uniform(0, 10), random.uniform(0, 10)) for _ in range(k)]

    # Assign each data point to the nearest centroid
    for _ in range(max_iters):
        assignments = [min(range(k), key=lambda j: euclidean_distance(point, centroids[j])) for point in data]

        # Update centroids based on the assigned points
        centroid_sums = [(0, 0) for _ in range(k)]
        counts = [0] * k

        for i, centroid_idx in enumerate(assignments):
            centroid_sums[centroid_idx] = (centroid_sums[centroid_idx][0] + data[i][0], centroid_sums[centroid_idx][1] + data[i][1])
            counts[centroid_idx] += 1

        centroids = [(centroid_sum[0] / count, centroid_sum[1] / count) if count > 0 else centroid for centroid, (centroid_sum, count) in zip(centroids, zip(centroid_sums, counts))]

    return assignments

def main():
    # Set the parameters for the dataset and clustering
    num_clusters = 3
    num_samples_train = 300
    num_samples_test = 100
    max_iters = 100

    # Generate random training dataset
    train_data = [(random.uniform(0, 10), random.uniform(0, 10)) for _ in range(num_samples_train)]

    # Perform K-means clustering on the training dataset
    train_assignments = kmeans(train_data, num_clusters, max_iters)

    # Get centroids directly from the training phase
    centroids = [(sum(point[0] for point, cluster in zip(train_data, train_assignments) if cluster == i) / max(1, train_assignments.count(i)),
                  sum(point[1] for point, cluster in zip(train_data, train_assignments) if cluster == i) / max(1, train_assignments.count(i)))
                 for i in range(num_clusters)]

    # Generate random test dataset
    test_data = [(random.uniform(0, 10), random.uniform(0, 10)) for _ in range(num_samples_test)]

    # Assign test samples to clusters using the centroids from training
    test_assignments = [min(range(num_clusters), key=lambda j: euclidean_distance(point, centroids[j])) for point in test_data]

    # Print accuracy
    correct_assignments = sum(1 for i, cluster in enumerate(test_assignments) if cluster == train_assignments[i])
    accuracy = correct_assignments / num_samples_test
    print(f"Accuracy: {accuracy:.2f}")

if __name__ == "__main__":
    main()

File: code-external/kmeans.rs
use rand::Rng;

// Euclidean distance function for K-means
fn euclidean_distance(a: &(f64, f64), b: &(f64, f64)) -> f64 {
    ((a.0 - b.0).powi(2) + (a.1 - b.1).powi(2)).sqrt()
}

// Perform K-means clustering
fn kmeans(data: &[(f64, f64)], k: usize, max_iters: usize) -> Vec<usize> {
    let mut rng = rand::thread_rng();

    // Initialize centroids randomly
    let mut centroids: Vec<(f64, f64)> = (0..k)
        .map(|_| (rng.gen_range(0.0..10.0), rng.gen_range(0.0..10.0)))
        .collect();

    let mut assignments: Vec<usize> = vec![0; data.len()];

    for _ in 0..max_iters {
        // Assign each data point to the nearest centroid
        for (i, point) in data.iter().enumerate() {
            let mut min_distance = f64::INFINITY;
            let mut closest_centroid = 0;

            for (j, centroid) in centroids.iter().enumerate() {
                let distance = euclidean_distance(point, centroid);

                if distance < min_distance {
                    min_distance = distance;
                    closest_centroid = j;
                }
            }

            assignments[i] = closest_centroid;
        }

        // Update centroids based on the assigned points
        let mut centroid_sums: Vec<(f64, f64)> = vec![(0.0, 0.0); k];
        let mut counts: Vec<usize> = vec![0; k];

        for (i, &centroid_idx) in assignments.iter().enumerate() {
            centroid_sums[centroid_idx].0 += data[i].0;
            centroid_sums[centroid_idx].1 += data[i].1;
            counts[centroid_idx] += 1;
        }

        for (i, count) in counts.iter().enumerate() {
            if *count > 0 {
                centroids[i].0 = centroid_sums[i].0 / *count as f64;
                centroids[i].1 = centroid_sums[i].1 / *count as f64;
            }
        }
    }

    assignments
}

fn main() {
    // Set the parameters for the dataset and clustering
    let num_clusters = 3;
    let num_samples_train = 300;
    let num_samples_test = 100;
    let max_iters = 100;

    // Generate random training dataset
    let mut rng = rand::thread_rng();
    let train_data: Vec<_> = (0..num_samples_train)
        .map(|_| (rng.gen_range(0.0..10.0), rng.gen_range(0.0..10.0)))
        .collect();

    // Perform K-means clustering on the training dataset
    let train_assignments = kmeans(&train_data, num_clusters, max_iters);

    // Get centroids directly from the training phase
    let centroids: Vec<_> = (0..num_clusters)
        .map(|cluster| {
            let cluster_points: Vec<_> = train_data
                .iter()
                .zip(train_assignments.iter())
                .filter(|(_, &c)| c == cluster)
                .map(|(point, _)| point)
                .collect();

            let sum: (f64, f64) = cluster_points.iter().cloned().fold((0.0, 0.0), |acc, p| {
                (acc.0 + p.0, acc.1 + p.1)
            });

            let count = cluster_points.len() as f64;
            (sum.0 / count, sum.1 / count)
        })
        .collect();

    // Generate random test dataset
    let test_data: Vec<_> = (0..num_samples_test)
        .map(|_| (rng.gen_range(0.0..10.0), rng.gen_range(0.0..10.0)))
        .collect();

    // Assign test samples to clusters using the centroids from training
    let test_assignments: Vec<usize> = test_data
        .iter()
        .map(|point| {
            let mut min_distance = f64::INFINITY;
            let mut closest_centroid = 0;

            for (j, centroid) in centroids.iter().enumerate() {
                let distance = euclidean_distance(point, centroid);

                if distance < min_distance {
                    min_distance = distance;
                    closest_centroid = j;
                }
            }

            closest_centroid
        })
        .collect();

    // Print accuracy
    let correct_assignments: usize = test_assignments
        .iter()
        .enumerate()
        .filter(|&(i, &cluster)| cluster == train_assignments[i])
        .count();

    let accuracy = correct_assignments as f64 / num_samples_test as f64;
    println!("Accuracy: {:.2}", accuracy);
}
